### PR TITLE
Make spanKindExtractor configurable in Ktor instrumentations

### PR DESCRIPTION
**Pull request by orangain**
This PR solves https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/8139. Now spanKindExtractor is configurable in Ktor instrumentations.

This make it possible to customize the SpanKind depending on the request as following:

```kt

    install(KtorServerTracing) {
        setOpenTelemetry(openTelemetry)
        setSpanKindExtractor {
            SpanKindExtractor { req ->
                if (req.uri.startsWith("/from-pubsub/")) {
                    SpanKind.CONSUMER
                } else {
                    SpanKind.SERVER
                }
            }
        }
    }
```

### CODE REVIEW

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java
Comment (orangain) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: I am not sure if it is ok to add such a public method to the InstrumenterBuilder, but I could not think of a better way to make SpanKindExtractor configurable.

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java
Comment (orangain) [inline instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:228]: This method is not used at this time, but I added it to make it symmetrical with the buildIncomingInstrumenter.

File: instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt
Comment (orangain) [inline instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt:59]: I have not written any tests for this method, as there were no unit tests for the existing setStatusExtractor or addAttributeExtractor, is that ok?

File: instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java
Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: Yeah, it is necessary to add a new method here (I kinda forgot about it earlier, when we were talking in GH discussions).

Since this is a stable public API, we need to make sure we come up with a good name. I'll list a few ideas I came up with here:
* `buildIncomingInstrumenter`/`buildOutgoingInstrumenter` (pretty much what this PR proposes)
* `buildContextExtractingInstrumenter`/`buildContextInjectingInstrumenter`
* `buildUpstreamInstrumenter`/`buildDownstreamInstrumeter`
* `buildPropagatingToUpstreamInstrumenter`/`buildPropagatingToDownstreamInstrumeter`
* `buildPropagatingInstrumenter` for both

@open-telemetry/java-instrumentation-maintainers WDYT?

Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: Naming is hard. Perhaps we should adopt a `@Incubating` or `@Experimental` annotations so such changes could be introduced more easily.
I think it is possible to implement this feature without introducing new api. You could use `buildInstrumenter(spanKindExtractor)` and do the context extraction manually before `instrumenter.start()` similarly to like it is done in `PropagatingFromUpstreamInstrumenter`.

Comment (orangain) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: @laurit  Thank you for your suggestion. I tried implementing this without introducing a new API.
https://github.com/orangain/opentelemetry-java-instrumentation/commit/daec120a46dca5c5581b0ed9880dec3e81ee65b4

While it is indeed possible to implement, I have a few concerns.

* Doing context propagation within library instrumentation seems to be a violation of responsibilities. Wouldn't it be confusing if only Ktor instrumentation does context propagation on its own, while `Instrumenter` does it in other library instrumentations?
* `ContextPropagationDebug` exists in the `internal` package, is it acceptable to use it in Ktor instrumentation? If not allowed, there is a concern that Ktor users will not be able to debug context propagation.

Comment (laurit) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: @orangain While letting `Instrumenter` handle context propagation is preferable there are instrumentations that do it manually for a variety of reasons. This was just a suggestion on how to move this forward faster, if you so wish you can ignore it. You could also start with manual propagation and when the methods you need are added to `Instrumenter` you can start using them.
Using internal classes is fine everywhere in this repository. Marking classes internal is more for informing users outside this repository that these classes could change without prior notice. I think that in this case not using `ContextPropagationDebug` would also be fine.

Comment (mateuszrzeszutek) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: Hey @orangain I opened #8392 to unblock you PR -- I'll let you know when it's merged so that you can rebase on top of it.

Comment (trask) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java:239]: hey @orangain, #8392 is merged now

Comment (orangain) [reply instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/InstrumenterBuilder.java]: @mateuszrzeszutek @trask @laurit  Sorry for the late response. Thank you for unblocking my PR! I've rebased the PR to use `InstrumenterUtil.buildUpstreamInstrumenter`.

File: instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt
Comment (orangain) [reply instrumentation/ktor/ktor-1.0/library/src/main/kotlin/io/opentelemetry/instrumentation/ktor/v1_0/KtorServerTracing.kt:60]: Thank you for your detailed explanation, I added test cases in https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/8255/commits/959ee02294c2244fa56a078ecd9d05c9292954c1.

**Review by mateuszrzeszutek — APPROVED** [review]
Thanks @orangain !
LGTM 👍

**Review by trask — APPROVED** [review]

**Pull request merged**
